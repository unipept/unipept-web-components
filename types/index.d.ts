export * from "./business";
export * from "./components";
export * from "./state";
