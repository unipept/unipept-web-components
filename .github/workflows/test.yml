name: Test

on: [push]

jobs:
  lint:
    runs-on: ubuntu-latest
    
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - name: Install dependencies
      run: npm install
    - name: Install Jest CLI
      run: npm i -g jest-cli
    - name: Run tests
      run: jest
    
