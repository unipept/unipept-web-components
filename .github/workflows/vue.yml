name: Vue

on: [push]

jobs:
  lint:
    runs-on: ubuntu-latest
    
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - name: Install dependencies
      run: npm install
    - name: Run ESLint
      run: eslint "src/**"
    
