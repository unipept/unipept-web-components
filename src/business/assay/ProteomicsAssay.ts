import Assay from "@/business/assay/Assay";
import SearchConfiguration from "@/business/configuration/SearchConfiguration";

export default class ProteomicsAssay extends Assay {
    constructor(
        protected readonly changeListeners: ChangeListener<Assay>[],
        public readonly id: string,
        public readonly peptides: string[],
        protected searchConfiguration: SearchConfiguration,
        protected name?: string,
        protected date?: Date,
    ) {
        super(changeListeners, id, name, date);
    }

    public getSearchConfiguration(): SearchConfiguration {
        return this.searchConfiguration;
    }

    public setSearchConfiguration(value: SearchConfiguration) {
        super.onUpdate("searchConfiguration", this.searchConfiguration, value);
        this.searchConfiguration = value;
    }

    async accept(visitor: AssayVisitor): Promise<void> {
        await visitor.visitMetaProteomicsAssay(this);
    }
}
