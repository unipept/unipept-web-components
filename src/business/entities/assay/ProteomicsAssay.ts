import Assay from "./Assay";
import SearchConfiguration from "./../../configuration/SearchConfiguration";
import AssayVisitor from "./AssayVisitor";
import { Peptide } from "./../../ontology/raw/Peptide";
import AnalysisSource from "./../../communication/analysis/AnalysisSource";
import { v4 as uuidv4 } from "uuid";

export default class ProteomicsAssay extends Assay {
    protected amountOfPeptides: number = 0;

    private searchConfiguration: SearchConfiguration = new SearchConfiguration();
    private peptides: Peptide[] = [];
    private analysisSource: AnalysisSource;

    constructor(
        public id: string = uuidv4()
    ) {
        super(id);
    }

    public getSearchConfiguration(): SearchConfiguration {
        return this.searchConfiguration;
    }

    public setSearchConfiguration(value: SearchConfiguration) {
        this.searchConfiguration = value;
    }

    public getPeptides(): Peptide[] {
        return this.peptides;
    }

    public setPeptides(peptides: Peptide[]): void {
        this.amountOfPeptides = peptides.length;
        this.peptides = peptides;
    }

    public getAmountOfPeptides(): number {
        return this.amountOfPeptides;
    }

    public setAmountOfPeptides(amount: number): void {
        this.amountOfPeptides = amount;
    }

    public getAnalysisSource(): AnalysisSource {
        return this.analysisSource;
    }

    public setAnalysisSource(source: AnalysisSource): void {
        this.analysisSource = source;
    }

    public accept(visitor: AssayVisitor): Promise<void> {
        return visitor.visitProteomicsAssay(this);
    }
}
