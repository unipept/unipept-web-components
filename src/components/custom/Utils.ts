export default class Utils {
    /**
     * This method should be used when a specific URL should be opened in a new browser window. The method automatically
     * decides whether Electron or a default redirection should take place.
     * 
     * @param url The full url to which navigation should take place.
     */
    static openInBrowser(url: string): void {
        if (Utils.isElectron()) {
            const shell = require("electron").shell;
            shell.openExternal(url);
        } else {
            window.open(url);
        }
    }
    
    /**
     * Method that can be used to check if the app is being executed by Electron or something else.
     * 
     * @return True if the app is being executed in an Electron-powered environment.
     */
    static isElectron(): boolean {
        const userAgent = navigator.userAgent.toLowerCase();
        return userAgent.indexOf(" electron/") > -1;
    }
}
