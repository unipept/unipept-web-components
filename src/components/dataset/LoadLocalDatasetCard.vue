<template>
    <v-card flat>
        <v-card-text v-if="storedDatasets.length === 0">
            <span>There are currently no datasets present in your browser's local storage.</span>
        </v-card-text>
        <v-list two-line>
            <template v-for="dataset of storedDatasets">
                <v-list-item :key="dataset.id" ripple @click="selectDataset(dataset)">
                    <v-list-item-action>
                        <tooltip message="Select this dataset for analysis.">
                            <v-icon>mdi-plus</v-icon>
                        </tooltip>
                    </v-list-item-action>
                    <v-list-item-content>
                        <v-list-item-title>
                            {{ dataset.getName() }}
                        </v-list-item-title>
                        <v-list-item-subtitle>
                            {{ dataset.getAmountOfPeptides() }} peptides
                        </v-list-item-subtitle>
                    </v-list-item-content>

                    <v-list-item-action>
                        <v-list-item-action-text>
                            {{ dataset.getDateFormatted() }}
                        </v-list-item-action-text>
                        <tooltip message="Delete this sample from local storage.">
                            <v-btn icon text @click="deleteDataset(dataset)" v-on:click.stop>
                                <v-icon color="grey darken-1">mdi-close</v-icon>
                            </v-btn>
                        </tooltip>
                    </v-list-item-action>
                </v-list-item>
            </template>
        </v-list>
    </v-card>
</template>

<script lang="ts">
import Vue from "vue";
import Component, { mixins } from "vue-class-component"
import Assay from "../../logic/data-management/assay/Assay";
import { Prop } from "vue-property-decorator";
import Tooltip from "../custom/Tooltip.vue";
import DatasetMixin from "./DatasetMixin.vue";

@Component({
    components: {
        Tooltip
    }
})
export default class LoadLocalDatasetCard extends mixins(DatasetMixin) {
    @Prop({ required: true })
    private storedDatasets: Assay[];
}
</script>

<style lang="less" scoped>
</style>
