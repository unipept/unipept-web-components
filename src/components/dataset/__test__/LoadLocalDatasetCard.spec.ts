import { shallowMount, createLocalVue, mount, Wrapper } from "@vue/test-utils";
import Vue from "vue";
import Vuetify from "vuetify";
import Vuex from "vuex";
import LoadLocalDatasetCard from "./../LoadLocalDatasetCard.vue";
import Setup from "@/test/Setup";
import Mock from "@/test/Mock";
import { sleep, waitForPromises, waitForElement, waitForCondition } from "@/test/Utils";
import { CombinedVueInstance } from "vue/types/vue";
import ProteomicsAssay from "@/business/entities/assay/ProteomicsAssay";


Vue.use(Vuetify);
Vue.use(Vuex);

const localVue = createLocalVue();

describe("LoadLocalDatasetCard", () => {
    let vuetify;
    let store;

    beforeEach(() => {
        vuetify = new Vuetify();
        store = new Vuex.Store({});
    });

    it("shows a placeholder if no datasets are present", async(done) => {
        const setup: Setup = new Setup();
        setup.setUpPrideNock();

        const wrapper = mount(LoadLocalDatasetCard, {
            store,
            localVue,
            vuetify,
            propsData: {
                storedAssays: []
            }
        });

        await waitForElement(wrapper, "#empty-dataset-placeholder");
        expect(wrapper.find("#empty-dataset-placeholder").html()).toMatchSnapshot();
        done();
    });

    it("shows a confirmation dialog before removing a dataset", async(done) => {
        const wrapper = await mountWithAssay(store, vuetify);

        await waitForElement(wrapper, ".remove-assay-button");
        wrapper.find(".remove-assay-button").trigger("click");

        await wrapper.vm.$nextTick();

        // Check if the dialog is active.
        expect(wrapper.find(".remove-confirmation-dialog .v-dialog").classes()).toContain("v-dialog--active");

        done();
    });

    it("correctly fires destroy-assay event after user confirmation", async(done) => {
        const wrapper = await mountWithAssay(store, vuetify);

        await waitForElement(wrapper, ".remove-assay-button");
        wrapper.find(".remove-assay-button").trigger("click");

        await wrapper.vm.$nextTick();
        wrapper.find(".confirmation-ok-button").trigger("click");

        await wrapper.vm.$nextTick();

        // Make sure that the destroy-assay event is fired
        expect(wrapper.emitted("destroy-assay")).toBeTruthy();

        done();
    });

    it("does not remove a dataset after clicking cancel on the confirmation", async(done) => {
        const wrapper = await mountWithAssay(store, vuetify);

        await waitForElement(wrapper, ".remove-assay-button");
        wrapper.find(".remove-assay-button").trigger("click");

        await wrapper.vm.$nextTick();
        wrapper.find(".confirmation-cancel-button").trigger("click");

        await wrapper.vm.$nextTick();

        // Make sure that the destroy-assay event is fired
        expect(wrapper.emitted("destroy-assay")).toBeFalsy();

        done();
    });

    it("correctly fires the create-assay event (and does not fire store-assay)", async(done) => {
        const wrapper = await mountWithAssay(store, vuetify);

        await waitForElement(wrapper, ".stored-assays-list .v-list-item");
        wrapper.find(".stored-assays-list .v-list-item").trigger("click");

        await wrapper.vm.$nextTick();

        expect(wrapper.emitted("create-assay")).toBeTruthy();
        expect(wrapper.emitted("store-assay")).toBeFalsy();

        done();
    });
});

const mountWithAssay = async function(store, vuetify): Promise<Wrapper<CombinedVueInstance<LoadLocalDatasetCard, object, object, object, Record<never, any>>>> {
    const setup: Setup = new Setup();
    setup.setUpPrideNock();

    const mock: Mock = new Mock();
    const assay: ProteomicsAssay = await mock.mockRealisticAssay();

    return mount(LoadLocalDatasetCard, {
        store,
        localVue,
        vuetify,
        propsData: {
            storedAssays: [assay]
        }
    })
}
