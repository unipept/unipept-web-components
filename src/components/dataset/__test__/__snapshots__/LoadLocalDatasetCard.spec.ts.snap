// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadLocalDatasetCard shows a placeholder if no datasets are present 1`] = `
"<span id=\\"empty-dataset-placeholder\\">
            There are currently no datasets present in your browser's local storage.
        </span>"
`;
